### PR TITLE
Implement Constraints on Subqueries     

### PR DESCRIPTION
**Pull request by dehowef**
Add logic to support constraints on subqueries.

Add an expression tree walker to traverse expr trees. This logic
can be expanded to search expr trees for other exprs outside of
subqueries as well.

Changed the grammar for COUNT () functions to use a keyword.
This avoids grammatical errors with other components.

Add regression tests for subqueries as well as some for the expr
tree walker.

### CODE REVIEW

**Review by jrgemignani — CHANGES_REQUESTED** [review]
Minor clarifications

File: src/backend/parser/cypher_analyze.c [hunk 1]
```diff
@@ -653,6 +660,381 @@ static int get_query_location(const int location, const char *source_str)
...
+    cr = makeNode(ColumnRef);
+    cr->fields = list_make1(makeNode(A_Star));
+    //cr->location = @1;
+
+    /*construct the restarget */
+    rt = makeNode(ResTarget);
+    rt->name = NULL;
+    rt->indirection = NIL;
+    rt->val = (Node *)cr;
+    //rt->location = @1;
```
[the remark this answers is not in the document]
Comment (dehowef) [reply src/backend/parser/cypher_analyze.c]: Thanks for catching that! I just pushed changes to set the location of the restarget and columnref!!!

**Review by jrgemignani — APPROVED** [review]

File: src/backend/parser/cypher_analyze.c [hunk 2]
```diff
@@ -679,8 +1061,26 @@ static Query *analyze_cypher(List *stmt, ParseState *parent_pstate,
         next->self = lfirst(lc);
         next->prev = clause;
 
+        /* check for subqueries in match */
+        if (is_ag_node(next->self, cypher_match))
+        {
+            cypher_match *match = (cypher_match *)next->self;
+
+             if (match->where != NULL && expr_contains_node(expr_has_subquery, match->where))
```
Comment (rafsun42) [inline src/backend/parser/cypher_analyze.c:1069]: Is the If-block is overindented?

File: src/backend/parser/cypher_clause.c [hunk 3]
```diff
@@ -448,6 +448,22 @@ static cypher_clause *make_cypher_clause(List *stmt)
         next->self = lfirst(lc);
         next->prev = clause;
 
+        /* check for subqueries in match */
+        if (is_ag_node(next->self, cypher_match))
+        {
+            cypher_match *match = (cypher_match *)next->self;
+
+             if (match->where != NULL && expr_contains_node(expr_has_subquery, match->where))
```
Comment (rafsun42) [inline src/backend/parser/cypher_clause.c:456]: Overindented if-block?

Comment (dehowef) [reply src/backend/parser/cypher_clause.c]: @rafsun42 Resolved

**Review by rafsun42 — APPROVED** [review]
Looks good!

**Pull request merged**